src/System.Text.Json/src/System/Text/Json/Serialization/ReadStackFrame.cs

```diff
@@ -58,9 +63,12 @@ internal struct ReadStackFrame
             !JsonPropertyInfo.IsPropertyPolicy &&
             JsonPropertyInfo.ClassType == ClassType.Enumerable;
 
-        public bool IsProcessingEnumerableOrDictionary => IsProcessingEnumerable || IsProcessingDictionary || IsProcessingImmutableDictionary;
+        public bool IsProcessingEnumerableOrDictionary => IsProcessingEnumerable || IsProcessingDictionary || IsProcessingIDictionaryConstructibleOrKeyValuePair;
+        public bool IsProcessingIDictionaryConstructibleOrKeyValuePair => IsProcessingIDictionaryConstructible || IsProcessingKeyValuePair;
```

Comment:
This new state and other recent state is becoming unwieldy. We need to clean this up in the future by using custom converters. This code is not going to be extensible otherwise and will probably be buggy.